File: 47. Permutations II.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums = sorted(nums)
        visited = [False for i in xrange(len(nums))]
        self.DFS(res, [], nums, visited)
        return res
        
    def DFS(self, res, path, nums, visited):
        if len(path) == len(nums):
            res += [path[:]]
            return
        for i in xrange(len(nums)):
            if i > 0 and nums[i] == nums[i-1] and not visited[i-1]:
                continue
            if not visited[i]:
                visited[i] = True
                self.DFS(res, path + [nums[i]], nums, visited)
                visited[i] = False

File: 47. Permutations II.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    let res = [];
    let visited = [];
    for (let i = 0; i < nums.length; i++)
        visited.push(false);
    nums = nums.sort((a,b) => a-b);
    var DFS = function(path, visited)
    {
        if (path.length == nums.length)
        {
            res.push([]);
            for (let i = 0; i < path.length; i++)
                res[res.length-1].push(path[i]);
            return;
        }
        for (let i = 0; i < nums.length; i++)
        {
            if (visited[i] || (i > 0 && nums[i] == nums[i-1] && !visited[i-1]))
                continue;
            visited[i] = true;
            path.push(nums[i]);
            DFS(path, visited);
            path.pop();
            visited[i] = false;
        }
    };
    DFS([], visited);
    return res;
};
